Up UEID to 256 bits and add design rational appendix (#35)
ietf-rats-wg/eat: draft-ietf-rats-eat.md

}
~~~~

# UEID Design Rationale

> ## Collision Probabilitty

Reviewer: Typo
```suggestion
## Collision Probability
```

UEIDs given, the total possible entity population and the number of
entities in a particular entity management database.

Three different sized databases are considered. The number of devices
per person roughly models non-personal devices such as traffic lights,
devices in stores they shop in, facilities they work in and so on,
even considering individual light bulbs. A device may have
individually attested subsystems, for example parts of a car or a
mobile phone. It is assumed that the largest database will have at
most 10% of the world's population of devices. Note that databases
that handle more than a trillion records exists today.

The trillion record database size models an easy-to-imagine reality
over the next decades. The quadrillion record database is a worst case
that might not come to pass, but would seem to be possible and thus
should be accommodated. The highly speculative 100 quadrillion case
perhaps involve nanorobots for every person and livestock animal and
domesticated bird. It is included to round out the analysis.

Note that the items counted here certainly do not have IP address and
are not individually connected to the network. They may be connected

> to internal buses, via serial links, blue tooth and so on.  This is

Reviewer 2: typo
```suggestion
to internal buses, via serial links, Bluetooth and so on.  This is
```

not the same problem as sizing IP addresses.

| People     | Devices / Person | Subsystems / Device | Database Portion | Database Size           |
|------------+------------------+------------------- -+------------------+-------------------------+
| 10 billion | 100              | 10                  | 1/10             | trillion (10^12)        | 
| 10 billion | 100,000          | 10                  | 1/10             | quadrillion (10^15)     | 
|100 billion | 1,000,000        | 10                  | 1/10             | 100 quadrillion (10^17) | 

This is conceptually similar to the Birthday Problem where m is the
number of possible birthdays, always 365, and k is the number of
people. It is also conceptually similar to the Birthday Attack where
collisions of the output of hash functions are considered.

The proper formula for the collision calculation is

       p = 1 - e^{-k^2/(2n)}
    
       p   Collision Probability
       n   Total possible population
       k   Actual population

> However, for the very large values involve here, this formula requires floating

Reviewer 2: ```suggestion
However, for the very large values involved here, this formula requires floating
```

> point precession higher than commonly available in calculators and SW so this

Reviewer 2: ```suggestion
point precision higher than commonly available in calculators and SW so this
```

Author: Grammar

The following tables gives the time interval until there is a probability of 
a collision based on there being one tenth the number of states per year
as the number of records in the database.
   
      t = 1 / ((k / 10) * p)
  
      t  Time until a collision
      p  Collision probability for UEID size
      k  Database size

| Database Size           | 128 bit UEID   | 192 bit UEID | 256 bit UEID |
|-------------------------+----------------+--------------+--------------+
| trillion (10^12)        | 60,000 years   | 10^24 years  | 10^44 years  |
| quadrillion (10^15)     | 8 seconds      | 10^14 years  | 10^34 years  |
| 100 quadrillion (10^17) | 8 microseconds | 10^11 years  | 10^31 years  |

Clearly 128 bits is not enough for the quadrillion-record database. Clearly
192 bits is enough for every case.

Note that even the rate of change of the database were much lower with one
thousandth of it changing per year, that would not improve the rate for
the one quadrillion record database to even one collision per day.  

The choice is for 256-bits to be prepared even for what can’t be
envisioned today. The table shows 192 bits is enough, but 256 bits,

> 32-bytes, is selected because it the output of the common SHA-256 hash

Reviewer 2: ```suggestion
32-bytes, is selected because it is the output of the common SHA-256 hash
```
